php_phongo.c

```diff
@@ -850,7 +850,24 @@ int phongo_execute_command(mongoc_client_t* client, php_phongo_command_type_t ty
 			return false;
 	}
 	if (!result) {
-		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+		if (error.domain == MONGOC_ERROR_SERVER || error.domain == MONGOC_ERROR_WRITE_CONCERN) {
+#if PHP_VERSION_ID >= 70000
+			zval zv;
+#else
+			zval* zv;
+#endif
+
+			zend_throw_exception(php_phongo_commandexception_ce, error.message, error.code TSRMLS_CC);
+			php_phongo_bson_to_zval(bson_get_data(&reply), reply.len, &zv);
+
+#if PHP_VERSION_ID >= 70000
+			phongo_add_exception_prop(ZEND_STRL("resultDocument"), &zv);
+#else
+			phongo_add_exception_prop(ZEND_STRL("resultDocument"), zv TSRMLS_CC);
+#endif
```

Comment:
Testing this locally, I found that it was still leaking. A debug build of PHP typically reports simple leaks on its own, which is how I caught it. Running the tests with Valgrind confirmed the same.

I don't expect that Homebrew uses debug builds of PHP, so no worries there.

Ultimately, I believe this is due to PHP adding a reference when it updates the property of an object. When `phongo_add_exception_prop()` uses [`zend_update_property()`](http://lxr.jpauli.tech/source/xref/php-7.2/Zend/zend_API.c#3953). Unless the object is some internal class that is also using a custom handler, this will call [`zend_std_write_property()`](http://lxr.jpauli.tech/source/xref/php-7.2/Zend/zend_object_handlers.c#683). In that function, you might notice that the zval gets a reference added to it.

I came up with https://github.com/kvwalker/mongo-php-driver/pull/1 while debugging this, although in hingsight it is possible that a simple `zval_ptr_dtor(&zv)` before the end of this scope might have fixed the leak. If you determine it does, feel free to do that and ignore my PR.

As for why this differs from `phongo_execute_bulk_write()` and it's assignment of `BulkWriteException::$writeResult`, I believe there are a few reasons:

 * There, we always initialize `return_value` as a WriteResult object, since it's either going to be used as the actual return value or stuffed into an exception. The only time we don't initialize it is if we know there isn't an error and the user is ignoring the return value.
 * When `phongo_add_exception_prop()` assigns `return_value` to the exception property, it gets a reference added. At that point, I expect its refcount is 2; however, PHP likely decrements the refcount on `return_value` when returning from `executeBulkWrite()` back to userland when an exception is thrown since it knows that userland has no way to access the function's return value.
 * In the event of an exception, we're left with a `return_value` with refcount=1 assigned to the exception object. It is ultimately freed along with the exception itself.